In [1]:
# 1. PostgreSQL
!apt-get -qq update
!apt-get -qq install -y postgresql postgresql-contrib
!service postgresql start
!sudo -u postgres psql -c "ALTER USER postgres PASSWORD 'postgres';"
!sudo -u postgres psql -c "CREATE DATABASE colab_db;" 2>/dev/null || true

W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu noble InRelease' does not seem to provide it (sources.list entry misspelt?)
[ OK ]
ALTER ROLE


In [2]:
# 2. SQL-магія (jupysql замість застарілого ipython-sql)
!pip uninstall -y -q ipython-sql
!pip install -q jupysql psycopg2-binary pandas

%load_ext sql
%config SqlMagic.displaycon = False
%config SqlMagic.feedback = False
%config SqlMagic.displaylimit = 20
%config SqlMagic.named_parameters = "disabled"
%sql postgresql+psycopg2://postgres:postgres@localhost:5432/colab_db

In [3]:
# 3. SQLAlchemy engine — потрібен для COPY FROM STDIN і pd.read_sql
from sqlalchemy import create_engine, text

engine = create_engine('postgresql+psycopg2://postgres:postgres@localhost:5432/colab_db')

# перевірка підключення: версія PostgreSQL
with engine.connect() as conn:
    print(conn.execute(text('SELECT version()')).scalar_one())

PostgreSQL 16.15 (Ubuntu 16.15-0ubuntu0.24.04.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0, 64-bit


In [4]:
import os
import pandas as pd
from urllib.request import urlretrieve

DATA_URL = 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2026-07.parquet'

os.makedirs('data', exist_ok=True)
parquet_path = 'data/hw3_taxi_month.parquet'
csv_path = 'data/hw3_taxi_sample.csv'

# Завантажуємо файл лише якщо його ще немає (щоб Restart & Run All не качав його повторно)
if not os.path.exists(parquet_path):
    urlretrieve(DATA_URL, parquet_path)

columns = [
    'VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
    'passenger_count', 'trip_distance', 'PULocationID', 'DOLocationID',
    'payment_type', 'fare_amount', 'tip_amount', 'total_amount',
    'store_and_fwd_flag'
]

df = pd.read_parquet(parquet_path, columns=columns).head(100_000)
df.to_csv(csv_path, index=False)

print(df.shape)
df.head()

(100000, 12)


,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,PULocationID,DOLocationID,payment_type,fare_amount,tip_amount,total_amount,store_and_fwd_flag
0,1,2026-07-01 00:36:25,2026-07-01 00:43:18,1.0,2.00,138,138,2,10.0,0.00,19.50,N
1,1,2026-07-01 00:02:53,2026-07-01 00:17:12,1.0,3.20,162,112,2,15.6,0.00,21.35,N
2,2,2026-07-01 00:18:56,2026-07-01 00:27:38,1.0,3.18,137,263,1,13.5,4.81,24.06,N
3,1,2026-07-01 00:35:06,2026-07-01 00:57:51,0.0,10.00,138,246,1,41.5,12.30,74.01,N
4,7,2026-07-01 00:36:59,2026-07-01 00:36:59,2.0,9.62,140,106,1,42.2,9.59,57.54,N


1.3. Staging-таблиця

In [5]:
%%sql

DROP TABLE IF EXISTS hw3_data_clean CASCADE;
DROP TABLE IF EXISTS hw3_data_staging CASCADE;

CREATE TABLE hw3_data_staging (
    vendor_id_raw TEXT,
    pickup_datetime_raw TEXT,
    dropoff_datetime_raw TEXT,
    passenger_count_raw TEXT,
    trip_distance_raw TEXT,
    pu_location_id_raw TEXT,
    do_location_id_raw TEXT,
    payment_type_raw TEXT,
    fare_amount_raw TEXT,
    tip_amount_raw TEXT,
    total_amount_raw TEXT,
    store_and_fwd_flag_raw TEXT
);

++
||
++
++

1.4. Завантаження через `COPY FROM STDIN`

In [6]:
raw_conn = engine.raw_connection()

try:
    with raw_conn.cursor() as cur:
        with open(csv_path, 'r', encoding='utf-8', newline='') as file:
            cur.copy_expert(
                """
                COPY hw3_data_staging (
                    vendor_id_raw,
                    pickup_datetime_raw,
                    dropoff_datetime_raw,
                    passenger_count_raw,
                    trip_distance_raw,
                    pu_location_id_raw,
                    do_location_id_raw,
                    payment_type_raw,
                    fare_amount_raw,
                    tip_amount_raw,
                    total_amount_raw,
                    store_and_fwd_flag_raw
                )
                FROM STDIN
                WITH (
                    FORMAT csv,
                    HEADER true,
                    DELIMITER ','
                )
                """,
                file
            )
    raw_conn.commit()
finally:
    raw_conn.close()

pd.read_sql('SELECT COUNT(*) AS loaded_rows FROM hw3_data_staging', engine)

,loaded_rows
0,100000


у staging завантажено 100 000 рядків, тобто стільки ж, скільки у сформованому CSV (`df.shape` = 100 000 × 12). Жоден рядок не загубився під час імпорту.

Завдання 2

In [7]:
%%sql

DROP TABLE IF EXISTS hw3_taxi_trips CASCADE;

CREATE TABLE hw3_taxi_trips (
    trip_id          BIGINT
        GENERATED BY DEFAULT AS IDENTITY
        PRIMARY KEY,

    vendor_id        SMALLINT,

    pickup_at        TIMESTAMPTZ NOT NULL,
    dropoff_at       TIMESTAMPTZ NOT NULL,

    pickup_date      DATE        NOT NULL,

    passenger_count  SMALLINT
        CHECK (passenger_count IS NULL OR passenger_count >= 0),

    trip_distance    NUMERIC(10, 2)
        CHECK (trip_distance >= 0),

    pu_location_id   INTEGER,
    do_location_id   INTEGER,

    payment_type     SMALLINT
        CHECK (payment_type IS NULL OR payment_type BETWEEN 0 AND 6),

    fare_amount      NUMERIC(10, 2),
    tip_amount       NUMERIC(10, 2),
    total_amount     NUMERIC(10, 2),

    store_and_fwd    BOOLEAN,

    CHECK (dropoff_at >= pickup_at)
);

++
||
++
++



У staging усі колонки мають тип `TEXT`, щоб імпорт не падав на порожніх або дивних значеннях. Для кожної колонки в clean-таблиці спочатку роблю `TRIM` (прибираю пробіли) і `NULLIF(..., '')` (порожній рядок стає `NULL`), а потім перетворюю в потрібний тип.

- **`trip_id`** — номер рядка. У файлі такої колонки немає, тому я додала `BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY`, щоб номер створювався автоматично.
- **`vendor_id`** — код компанії, яка передала запис. У CSV це ціле число (`1`, `2`, ...). Тип `SMALLINT`, бо це невеликий код, а не число для розрахунків.
- **`pickup_at`, `dropoff_at`** — час посадки і висадки. У CSV це текст типу `2026-07-01 00:36:25` без часового поясу (це час Нью-Йорка). Тип `TIMESTAMPTZ`, бо це конкретний момент часу, тому додаю пояс через `AT TIME ZONE 'America/New_York'`. `NOT NULL`, бо без часу запис про поїздку не має сенсу.
- **`pickup_date`** — дата посадки без часу. Беру її з `pickup_datetime_raw` через `::DATE`. Тип `DATE`, бо тут потрібна тільки дата — зручно групувати по днях.
- **`passenger_count`** — кількість пасажирів. У CSV вона записана як `1.0`, `2.0`, тому спочатку перетворюю в `NUMERIC`, а потім у `SMALLINT`. Додала `CHECK (passenger_count >= 0)`, бо від'ємної кількості пасажирів не буває.
- **`trip_distance`** — відстань поїздки в милях, наприклад `2.38`. Тип `NUMERIC(10,2)` — два знаки після коми. `CHECK (trip_distance >= 0)`, бо відстань не може бути від'ємною.
- **`pu_location_id`, `do_location_id`** — номери зон посадки і висадки. У CSV це цілі числа. Тип `INTEGER`.
- **`payment_type`** — код способу оплати (1 — картка, 2 — готівка тощо). Тип `SMALLINT`. `CHECK (payment_type BETWEEN 0 AND 6)` — за описом даних TLC бувають лише коди від 0 до 6.
- **`fare_amount`, `tip_amount`, `total_amount`** — тариф, чайові і загальна сума в доларах, наприклад `13.50`. Тип `NUMERIC(10,2)`, бо це гроші з точністю до центів. `CHECK >= 0` тут не ставила, бо не хотіла втратити рядки з від'ємними сумами — вони перевіряються в пункті 4.3.
- **`store_and_fwd`** — ознака так/ні. У CSV там `Y` або `N`. Тип `BOOLEAN`: `Y` → `TRUE`, `N` → `FALSE`, інше → `NULL`.

`VARCHAR` не використовувала, бо текстових колонок у фінальній таблиці немає.

Перенесення даних staging у clean

In [8]:
%%sql
WITH parsed AS (
    SELECT
        NULLIF(TRIM(vendor_id_raw), '')::SMALLINT AS vendor_id,

        NULLIF(TRIM(pickup_datetime_raw), '')::TIMESTAMP
            AT TIME ZONE 'America/New_York' AS pickup_at,

        NULLIF(TRIM(dropoff_datetime_raw), '')::TIMESTAMP
            AT TIME ZONE 'America/New_York' AS dropoff_at,

        NULLIF(TRIM(pickup_datetime_raw), '')::TIMESTAMP::DATE AS pickup_date,

        NULLIF(TRIM(passenger_count_raw), '')::NUMERIC::SMALLINT AS passenger_count,

        NULLIF(TRIM(trip_distance_raw), '')::NUMERIC(10, 2) AS trip_distance,

        NULLIF(TRIM(pu_location_id_raw), '')::INTEGER AS pu_location_id,
        NULLIF(TRIM(do_location_id_raw), '')::INTEGER AS do_location_id,

        NULLIF(TRIM(payment_type_raw), '')::SMALLINT AS payment_type,

        NULLIF(TRIM(fare_amount_raw), '')::NUMERIC(10, 2) AS fare_amount,
        NULLIF(TRIM(tip_amount_raw), '')::NUMERIC(10, 2) AS tip_amount,
        NULLIF(TRIM(total_amount_raw), '')::NUMERIC(10, 2) AS total_amount,

        CASE UPPER(TRIM(store_and_fwd_flag_raw))
            WHEN 'Y' THEN TRUE
            WHEN 'N' THEN FALSE
        END AS store_and_fwd

    FROM hw3_data_staging
)
INSERT INTO hw3_taxi_trips (
    vendor_id, pickup_at, dropoff_at, pickup_date, passenger_count, trip_distance,
    pu_location_id, do_location_id, payment_type,
    fare_amount, tip_amount, total_amount, store_and_fwd
)
SELECT
    vendor_id, pickup_at, dropoff_at, pickup_date, passenger_count, trip_distance,
    pu_location_id, do_location_id, payment_type,
    fare_amount, tip_amount, total_amount, store_and_fwd
FROM parsed
WHERE pickup_at IS NOT NULL
  AND dropoff_at IS NOT NULL
  AND dropoff_at >= pickup_at
  AND trip_distance >= 0
  AND (passenger_count IS NULL OR passenger_count >= 0)
  AND (payment_type IS NULL OR payment_type BETWEEN 0 AND 6);

++
||
++
++

перевірка

In [9]:
%%sql
SELECT COUNT(*) AS loaded_trips FROM hw3_taxi_trips;

loaded_trips
100000


In [10]:
%%sql
SELECT * FROM hw3_taxi_trips LIMIT 5;

trip_id,vendor_id,pickup_at,dropoff_at,pickup_date,passenger_count,trip_distance,pu_location_id,do_location_id,payment_type,fare_amount,tip_amount,total_amount,store_and_fwd
1,1,2026-07-01 04:36:25+00:00,2026-07-01 04:43:18+00:00,2026-07-01,1,2.00,138,138,2,10.00,0.00,19.50,False
2,1,2026-07-01 04:02:53+00:00,2026-07-01 04:17:12+00:00,2026-07-01,1,3.20,162,112,2,15.60,0.00,21.35,False
3,2,2026-07-01 04:18:56+00:00,2026-07-01 04:27:38+00:00,2026-07-01,1,3.18,137,263,1,13.50,4.81,24.06,False
4,1,2026-07-01 04:35:06+00:00,2026-07-01 04:57:51+00:00,2026-07-01,0,10.00,138,246,1,41.50,12.30,74.01,False
5,7,2026-07-01 04:36:59+00:00,2026-07-01 04:36:59+00:00,2026-07-01,2,9.62,140,106,1,42.20,9.59,57.54,False


завдання 3

3.1. Контроль кількості рядків

In [11]:
%%sql
SELECT 'staging' AS layer, COUNT(*) AS n_rows
FROM hw3_data_staging

UNION ALL

SELECT 'clean' AS layer, COUNT(*) AS n_rows
FROM hw3_taxi_trips

layer,n_rows
staging,100000
clean,100000


у staging і clean однаково — по 100 000 рядків.

Рядки могли відфільтруватися лише на кроці 2.2, де дані переносяться з staging у clean (`INSERT INTO hw3_taxi_trips ... SELECT ... WHERE ...`). У `WHERE` там стоять умови: є час посадки і висадки, висадка не раніше посадки, відстань і кількість пасажирів не від'ємні, код оплати від 0 до 6. Кількість рядків не змінилась, отже всі 100 000 рядків ці умови пройшли і жоден не був відкинутий.

3.2. Перевірка дублікатів

Окремого id поїздки у файлі немає, тому як business key беру комбінацію: вендор, час посадки і висадки, зони посадки і висадки, відстань і суми.

In [12]:
%%sql
SELECT
    COUNT(*) AS total_rows,
    COUNT(DISTINCT (vendor_id, pickup_at, dropoff_at, pu_location_id, do_location_id,
                    trip_distance, fare_amount, total_amount)) AS distinct_strict_key,
    COUNT(*) - COUNT(DISTINCT (vendor_id, pickup_at, dropoff_at, pu_location_id, do_location_id,
                               trip_distance, fare_amount, total_amount)) AS strict_duplicate_rows
FROM hw3_taxi_trips;

total_rows,distinct_strict_key,strict_duplicate_rows
100000,100000,0


за таким ключем повних дублікатів **0**, отже копій записів у вибірці немає.

3.3. Null-rate і спеціальні маркери

In [13]:
%%sql
SELECT
    ROUND(COUNT(*) FILTER (WHERE vendor_id IS NULL)       * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_vendor,
    ROUND(COUNT(*) FILTER (WHERE pickup_at IS NULL)       * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_pickup,
    ROUND(COUNT(*) FILTER (WHERE dropoff_at IS NULL)      * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_dropoff,
    ROUND(COUNT(*) FILTER (WHERE passenger_count IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_passengers,
    ROUND(COUNT(*) FILTER (WHERE trip_distance IS NULL)   * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_distance,
    ROUND(COUNT(*) FILTER (WHERE payment_type IS NULL)    * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_payment,
    ROUND(COUNT(*) FILTER (WHERE fare_amount IS NULL)     * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_fare_amount,
    ROUND(COUNT(*) FILTER (WHERE total_amount IS NULL)    * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_total_amount,
    ROUND(COUNT(*) FILTER (WHERE store_and_fwd IS NULL)   * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_store_fwd,
    COUNT(*) AS total_rows
FROM hw3_taxi_trips;

pct_null_vendor,pct_null_pickup,pct_null_dropoff,pct_null_passengers,pct_null_distance,pct_null_payment,pct_null_fare_amount,pct_null_total_amount,pct_null_store_fwd,total_rows
0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,100000


у всіх перевірених колонках `NULL` немає (0.00 %). Але це ще не означає, що пропусків немає зовсім — пропуск може бути записаний як порожній рядок або як `0`. Тому перевіряю сирі дані з staging:

In [14]:
%%sql
SELECT
    COUNT(*) FILTER (WHERE NULLIF(TRIM(passenger_count_raw), '') IS NULL)          AS empty_passenger_count,
    COUNT(*) FILTER (WHERE NULLIF(TRIM(store_and_fwd_flag_raw), '') IS NULL)       AS empty_store_fwd_flag,
    COUNT(*) FILTER (WHERE TRIM(passenger_count_raw) IN ('0', '0.0'))              AS zero_passengers,
    COUNT(*) FILTER (WHERE NULLIF(TRIM(trip_distance_raw), '')::NUMERIC = 0)       AS zero_distance,
    COUNT(*) FILTER (WHERE TRIM(pickup_datetime_raw) = TRIM(dropoff_datetime_raw)) AS zero_duration,
    COUNT(*) AS total_rows
FROM hw3_data_staging;

empty_passenger_count,empty_store_fwd_flag,zero_passengers,zero_distance,zero_duration,total_rows
0,0,312,1365,1800,100000


**Висновок:** порожніх рядків немає. Але є поїздки з `0` пасажирів (312), з нульовою відстанню (1 365) і з однаковим часом посадки та висадки (1 800). Поїздка не може бути без пасажирів, на 0 миль чи на 0 хвилин, тому, найімовірніше, це пропущені або неправильно записані значення. Я їх не видаляю, але при аналізі їх варто враховувати окремо.

Завдання 4

4.1. Розподіл категоріальної колонки

In [15]:
%%sql
SELECT
    payment_type,
    COUNT(*) AS n_rows,
    ROUND(
        COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (),
        2
    ) AS pct_rows
FROM hw3_taxi_trips
GROUP BY payment_type
ORDER BY n_rows DESC, payment_type

payment_type,n_rows,pct_rows
1,85396,85.40
2,13384,13.38
4,777,0.78
3,443,0.44


**Висновок:** найчастіше платять карткою (код 1) — **85.40 %** поїздок, готівкою (код 2) — **13.38 %**. Коди 4 (спір, 0.78 %) і 3 (без оплати, 0.44 %) трапляються дуже рідко. Кодів 0, 5 і 6 у вибірці немає.
Розподіл дуже нерівномірний: майже всі поїздки оплачені карткою. Якщо будувати модель, рідкісні коди 3 і 4 вона майже не «побачить», тому їх варто аналізувати окремо.

4.2. Базова статистика числової колонки

In [16]:
%%sql
SELECT
    COUNT(total_amount) AS n_non_null,
    MIN(total_amount) AS min_value,
    MAX(total_amount) AS max_value,
    ROUND(AVG(total_amount)::NUMERIC, 2) AS avg_value,
    ROUND(STDDEV_SAMP(total_amount)::NUMERIC, 2) AS sd_value
FROM hw3_taxi_trips

n_non_null,min_value,max_value,avg_value,sd_value
100000,-801.00,801.00,28.07,23.40


**Висновок:** усі 100 000 значень `total_amount` заповнені. Середня сума поїздки — **$28.07**, стандартне відхилення — **$23.40**, тобто суми дуже різні: є багато дешевих поїздок і небагато дорогих.
Дивно, що мінімум **−801.00**, а максимум **801.00** — однакові числа з різними знаками. Від'ємна сума за поїздку неможлива, тому, схоже, це повернення грошей за ту саму поїздку. Такі рядки я перевіряю в 4.3.

4.3. Один outlier-check

domain-specific перевірки

Перевіряємо значення, неможливі для звичайної оплаченої поїздки: від'ємний тариф, від'ємні чайові або нереалістично велика сума (> $10 000).

In [17]:
%%sql
SELECT *
FROM hw3_taxi_trips
WHERE fare_amount < 0
   OR tip_amount < 0
   OR total_amount > 10000
ORDER BY pickup_at
LIMIT 20;

trip_id,vendor_id,pickup_at,dropoff_at,pickup_date,passenger_count,trip_distance,pu_location_id,do_location_id,payment_type,fare_amount,tip_amount,total_amount,store_and_fwd
246,2,2026-07-01 04:00:07+00:00,2026-07-01 04:09:49+00:00,2026-07-01,1,2.38,237,226,4,-13.50,0.00,-19.25,False
1230,2,2026-07-01 04:01:47+00:00,2026-07-01 04:22:50+00:00,2026-07-01,1,8.26,161,49,2,-35.20,0.00,-48.41,False
435,2,2026-07-01 04:08:35+00:00,2026-07-01 04:24:37+00:00,2026-07-01,1,4.56,209,190,4,-22.60,0.00,-28.35,False
393,2,2026-07-01 04:13:43+00:00,2026-07-01 04:15:59+00:00,2026-07-01,1,0.00,138,138,4,-4.40,0.00,-13.90,False
1377,2,2026-07-01 04:22:26+00:00,2026-07-01 05:10:32+00:00,2026-07-01,2,18.65,198,174,2,-98.60,0.00,-110.35,False
552,2,2026-07-01 04:23:39+00:00,2026-07-01 04:23:49+00:00,2026-07-01,4,0.00,79,79,2,-12.00,0.00,-16.25,False
233,2,2026-07-01 04:32:01+00:00,2026-07-01 04:41:58+00:00,2026-07-01,1,1.80,68,230,4,-11.40,0.00,-17.15,False
1313,2,2026-07-01 04:35:36+00:00,2026-07-01 05:13:59+00:00,2026-07-01,1,16.01,132,129,3,-87.40,0.00,-91.40,False
1305,2,2026-07-01 04:36:34+00:00,2026-07-01 04:43:04+00:00,2026-07-01,1,1.09,100,170,3,-8.60,0.00,-14.35,False
1467,2,2026-07-01 04:41:01+00:00,2026-07-01 05:13:11+00:00,2026-07-01,1,18.41,132,55,4,-69.50,0.00,-74.00,False


помітила що такі поїздки у вендор ид = 2 хочу перевірити чи це дійсно так

In [18]:
%%sql
SELECT
    vendor_id,
    COUNT(*) AS n_trips
FROM hw3_taxi_trips
WHERE fare_amount < 0
   OR tip_amount < 0
   OR total_amount > 10000
GROUP BY vendor_id
ORDER BY n_trips DESC;

vendor_id,n_trips
2,493


поїздок із сумою понад $10 000 немає, але є **493** записи з від'ємними сумами, і всі вони від `vendor_id = 2` (можливо це якийсь тестовий варіант або інша можлива причина)

Я б їх не видаляла, а:
- додала б прапорець (наприклад, `is_negative_amount`), щоб відрізняти такі рядки від звичайних поїздок;
- виключила б їх з training data, якщо модель прогнозує вартість поїздки;
- передала б на перевірку джерела — уточнила б у вендора 2, що це за записи.

Завдання 5.

In [19]:
%%sql
-- 5.1 Чи є в сирих даних поїздки, у яких дата посадки не з липня 2026?
SELECT
    LEFT(pickup_datetime_raw, 7)  AS pickup_month_raw,
    COUNT(*)                      AS n_rows,
    MIN(pickup_datetime_raw)      AS min_pickup_raw,
    MAX(pickup_datetime_raw)      AS max_pickup_raw
FROM hw3_data_staging
WHERE pickup_datetime_raw NOT LIKE '2026-07-%%'
GROUP BY pickup_month_raw
ORDER BY n_rows DESC;

pickup_month_raw,n_rows,min_pickup_raw,max_pickup_raw
2026-06,7,2026-06-30 23:39:59,2026-06-30 23:59:58


Так, знайшлося **7** поїздок з датою посадки 30 червня 2026 (між 23:39 і 23:59). Це поїздки, які почалися в останні хвилини червня і, мабуть, закінчилися вже в липні, тому потрапили в липневий файл. Їх небагато, але для аналізу по днях варто пам'ятати, що у файлі можуть бути дати з іншого місяця.

In [20]:
%%sql
-- 5.2 Скільки поїздок починається і закінчується в тій самій зоні?
SELECT
    CASE
        WHEN pu_location_id IS DISTINCT FROM do_location_id THEN 'різні зони'
        ELSE 'та сама зона'
    END                                                            AS zone_relation,
    COUNT(*)                                                       AS n_trips,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2)             AS pct_trips,
    ROUND(AVG(trip_distance), 2)                                   AS avg_distance_mi,
    ROUND(AVG(EXTRACT(EPOCH FROM dropoff_at - pickup_at) / 60), 1) AS avg_duration_min
FROM hw3_taxi_trips
GROUP BY zone_relation
ORDER BY n_trips DESC;

zone_relation,n_trips,pct_trips,avg_distance_mi,avg_duration_min
різні зони,94298,94.30,3.15,16.0
та сама зона,5702,5.70,0.63,5.0


Більшість поїздок (**94.30 %**) — між різними зонами, і лише **5.70 %** починаються і закінчуються в тій самій зоні. Поїздки в межах однієї зони набагато коротші: в середньому **0.63 милі** і **5 хв** проти **3.15 милі** і **16 хв** для різних зон. Тобто це окрема група дуже коротких поїздок.

In [21]:
%%sql
-- 5.3 Які унікальні комбінації вендора і способу оплати є в даних?
SELECT DISTINCT
    vendor_id,
    payment_type
FROM hw3_taxi_trips
ORDER BY vendor_id, payment_type;

vendor_id,payment_type
1,1
1,2
1,3
1,4
2,1
2,2
2,3
2,4
7,1
7,2


У вибірці є три вендори — **1, 2 і 7**, і в кожного трапляються ті самі коди оплати **1–4**. Кодів 0, 5 і 6 немає в жодного вендора. Отже, всі вендори заповнюють `payment_type` однаковим набором кодів.

In [22]:
%%sql
-- 5.4 У які години доби (за часом Нью-Йорка) найбільше поїздок?
SELECT
    EXTRACT(HOUR FROM pickup_at AT TIME ZONE 'America/New_York')::INT AS pickup_hour,
    COUNT(*)                                                          AS n_trips,
    ROUND(AVG(total_amount), 2)                                       AS avg_total_amount
FROM hw3_taxi_trips
GROUP BY pickup_hour
ORDER BY n_trips DESC, pickup_hour
LIMIT 5;

pickup_hour,n_trips,avg_total_amount
9,6855,25.09
8,6719,25.49
18,6614,28.56
17,6441,29.55
16,5953,31.06


** Найбільше поїздок о **9:00 (6 855)** і **8:00 (6 719)** ранку, а також о **18:00, 17:00 і 16:00** вечора. Це схоже на ранкову і вечірню годину пік, коли люди їдуть на роботу і з роботи. Увечері середня сума поїздки вища (**$28–31**), ніж уранці (**~$25**).

In [23]:
%%sql
-- 5.5 Які дні тижня найпопулярніші для поїздок?
SELECT
    EXTRACT(ISODOW FROM pickup_date)::INT AS day_num,
    TO_CHAR(pickup_date, 'Day')           AS day_name,
    COUNT(*)                              AS n_trips,
    COUNT(DISTINCT pickup_date)           AS n_dates_in_sample
FROM hw3_taxi_trips
GROUP BY day_num, day_name
ORDER BY n_trips DESC;

day_num,day_name,n_trips,n_dates_in_sample
3,Wednesday,86535,1
4,Thursday,13458,1
2,Tuesday,7,1


** У вибірці є лише **3 дати**: вівторок (7 поїздок), середа (**86 535**) і четвер (**13 458**) — у стовпці `n_dates_in_sample` для кожного дня стоїть 1. Тобто перші 100 000 рядків — це майже повністю 1 липня. Порівнювати дні тижня на такій вибірці не можна, для цього потрібні дані за весь місяць.

In [24]:
%%sql
-- 5.6 Яка частка поїздок має невідому (NULL) або нульову кількість пасажирів?
SELECT
    COALESCE(passenger_count::TEXT, 'невідомо (NULL)')    AS passenger_label,
    CASE
        WHEN passenger_count IS NULL OR passenger_count = 0 THEN 'пропуск'
        ELSE 'є значення'
    END                                                   AS is_missing,
    COUNT(*)                                              AS n_trips,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2)    AS pct_trips
FROM hw3_taxi_trips
GROUP BY passenger_count
ORDER BY passenger_count NULLS FIRST;

passenger_label,is_missing,n_trips,pct_trips
0,пропуск,312,0.31
1,є значення,84488,84.49
2,є значення,10647,10.65
3,є значення,2451,2.45
4,є значення,1761,1.76
5,є значення,198,0.20
6,є значення,143,0.14


`NULL` у кількості пасажирів немає, але є **312** поїздок (**0.31 %**) з `0` пасажирів — це фактично пропуск (у колонці `is_missing` вони позначені як «пропуск»). Найчастіше їде **1 пасажир (84.49 %)**, 2 пасажири — 10.65 %, а 5–6 пасажирів — зовсім рідко.

In [25]:
%%sql
-- 5.7 Які 10 зон посадки найпопулярніші?
SELECT
    pu_location_id,
    COUNT(*)                                             AS n_trips,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2)   AS pct_trips,
    ROUND(AVG(total_amount), 2)                          AS avg_total_amount
FROM hw3_taxi_trips
GROUP BY pu_location_id
ORDER BY n_trips DESC, pu_location_id
LIMIT 10;

pu_location_id,n_trips,pct_trips,avg_total_amount
237,5645,5.65,20.21
161,5590,5.59,26.39
132,4911,4.91,76.98
236,4555,4.56,20.39
162,4187,4.19,24.44
230,3670,3.67,29.69
186,3532,3.53,26.21
142,3269,3.27,21.97
170,3106,3.11,23.90
163,3021,3.02,26.05


Найпопулярніші зони посадки — **237 (5.65 %)**, **161 (5.59 %)** і **132 (4.91 %)**. Кожна зона з топ-10 дає лише 3–6 % поїздок, тобто поїздки досить рівномірно розподілені між зонами. Помітно виділяється зона **132**: середня сума там **$76.98**, а в інших зонах лише **$20–30** — звідти, схоже, їздять набагато далі.

In [26]:
%%sql
-- 5.8 Який середній відсоток чайових для кожного типу оплати залежно від відстані поїздки?
SELECT
    payment_type,
    CASE
        WHEN trip_distance < 1                 THEN '1) до 1 милі'
        WHEN trip_distance BETWEEN 1 AND 2.99  THEN '2) 1-3 милі'
        WHEN trip_distance BETWEEN 3 AND 9.99  THEN '3) 3-10 миль'
        ELSE                                        '4) 10+ миль'
    END                                                        AS distance_group,
    COUNT(*)                                                   AS n_trips,
    ROUND(AVG(tip_amount * 100.0 / NULLIF(fare_amount, 0)), 2) AS avg_tip_pct
FROM hw3_taxi_trips
WHERE fare_amount > 0
GROUP BY payment_type, distance_group
ORDER BY payment_type, distance_group;

payment_type,distance_group,n_trips,avg_tip_pct
1,1) до 1 милі,23540,32.50
1,2) 1-3 милі,42161,25.41
1,3) 3-10 миль,14166,21.72
1,4) 10+ миль,5505,20.19
2,1) до 1 милі,4294,0.01
2,2) 1-3 милі,5257,0.00
2,3) 3-10 миль,2414,0.00
2,4) 10+ миль,1273,0.00
3,1) до 1 милі,220,0.00
3,2) 1-3 милі,85,0.00


Чайові є тільки при оплаті карткою (код 1): від **32.50 %** для поїздок до 1 милі до **20.19 %** для поїздок понад 10 миль. Для готівки (код 2) і кодів 3 і 4 чайові майже **0 %**, тобто в даних вони просто не записуються. Отже, аналізувати чайові можна лише на поїздках, оплачених карткою.

In [27]:
%%sql
-- 5.9 Які 10 найдорожчих поїздок у вибірці і звідки вони були?
SELECT
    trip_id,
    pickup_at,
    pu_location_id,
    do_location_id,
    trip_distance,
    payment_type,
    fare_amount,
    tip_amount,
    total_amount
FROM hw3_taxi_trips
WHERE (fare_amount > 0 AND total_amount > 0)
  AND NOT (payment_type IN (3, 4))
ORDER BY total_amount DESC
LIMIT 10;

trip_id,pickup_at,pu_location_id,do_location_id,trip_distance,payment_type,fare_amount,tip_amount,total_amount
33856,2026-07-01 17:35:32+00:00,132,265,78.93,2,511.20,0.00,536.45
70110,2026-07-01 23:55:53+00:00,132,265,83.58,2,484.60,0.00,527.81
46562,2026-07-01 20:33:26+00:00,132,265,81.17,2,467.80,0.00,505.01
64237,2026-07-01 22:44:14+00:00,132,265,77.52,2,487.40,0.00,500.36
96778,2026-07-02 12:38:56+00:00,132,265,79.86,2,440.50,0.00,475.21
49982,2026-07-01 20:52:23+00:00,132,265,84.90,1,450.00,20.00,473.00
38235,2026-07-01 18:39:22+00:00,132,265,69.48,1,438.40,0.00,449.36
86644,2026-07-02 04:10:23+00:00,133,133,0.00,2,447.50,0.00,448.50
33876,2026-07-01 17:24:03+00:00,230,230,0.00,1,430.00,0.00,434.25
2928,2026-07-01 08:09:17+00:00,230,265,49.27,1,339.70,71.45,428.69


Найдорожчі поїздки коштують **$430–536**. Більшість із них — із зони **132** у зону **265**, відстань **70–85 миль**, тож висока ціна тут виглядає реальною. Але є дві дивні поїздки (зони 133 → 133 і 230 → 230) з відстанню **0.00 миль** і сумою понад **$430** — їх варто перевірити окремо.

In [28]:
%%sql
-- 5.10 Скільки поїздок припадає на різні групи тривалості поїздки?
WITH durations AS (
    SELECT
        EXTRACT(EPOCH FROM dropoff_at - pickup_at) / 60 AS duration_min
    FROM hw3_taxi_trips
)
SELECT
    CASE
        WHEN duration_min = 0                  THEN '0) 0 хв'
        WHEN duration_min < 5                  THEN '1) до 5 хв'
        WHEN duration_min BETWEEN 5 AND 15     THEN '2) 5-15 хв'
        WHEN duration_min BETWEEN 15 AND 30    THEN '3) 15-30 хв'
        WHEN duration_min BETWEEN 30 AND 60    THEN '4) 30-60 хв'
        WHEN duration_min BETWEEN 60 AND 180   THEN '5) 1-3 год'
        ELSE                                        '6) більше 3 год'
    END                                                  AS duration_group,
    COUNT(*)                                             AS n_trips,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2)   AS pct_trips
FROM durations
GROUP BY duration_group
ORDER BY duration_group;

duration_group,n_trips,pct_trips
0) 0 хв,1800,1.80
1) до 5 хв,12529,12.53
2) 5-15 хв,49031,49.03
3) 15-30 хв,26645,26.65
4) 30-60 хв,8778,8.78
5) 1-3 год,1176,1.18
6) більше 3 год,41,0.04


Майже половина поїздок (**49.03 %**) триває **5–15 хв**, ще **26.65 %** — 15–30 хв. Є **1 800** поїздок (**1.80 %**) тривалістю 0 хв і **41** поїздка довше 3 годин — це підозрілі значення, які варто позначити, а не вважати звичайними поїздками.

Завдання 6

## Завдання 6. Reflection

**Проблеми якості.** Явних `NULL` у даних немає, але частина пропусків записана як `0`: поїздки з 0 пасажирів, з нульовою відстанню і з нульовою тривалістю. Найбільша проблема — 493 записи з від'ємними сумами, усі від `vendor_id = 2`; мінімум −801 і максимум 801 схожі на повернення коштів за ту саму поїздку. Дублікатів за business key не знайдено, але без id поїздки це не можна гарантувати. Крім того, я взяла перші 100 000 рядків, тому вибірка — це майже лише 1 липня, а ще в ній є 7 поїздок з 30 червня.

**Features.** Використала б годину й день посадки, зони посадки і висадки (`pu_location_id`, `do_location_id`), `vendor_id` і `passenger_count`. Не використала б `trip_id`, бо це просто номер рядка, і `store_and_fwd`, бо ця ознака стосується передачі даних, а не самої поїздки.

**Leakage.** Якщо прогнозувати вартість або чайові в момент посадки, то `dropoff_at`, тривалість, `trip_distance`, `tip_amount`, `total_amount` і `payment_type` стають відомі лише після завершення поїздки, тому їх не можна брати як ознаки. Особливо `total_amount` — він уже містить у собі тариф і чайові.

**Пропозиції для pipeline.** Додати унікальний id поїздки; позначати повернення окремим полем замість від'ємних сум; записувати час разом із часовим поясом; писати `NULL` замість `0`, якщо кількість пасажирів невідома; перевіряти, що дата поїздки належить місяцю файлу.

**ML-задача.** Регресія (прогноз вартості або тривалості поїздки), класифікація (чи залишить пасажир чайові) або forecasting (прогноз кількості поїздок за годинами і зонами).